### PR TITLE
[Feature] split scan or predicate to union

### PR DESCRIPTION
**Pull request by packy92**
Fixes #issue
### Main work
For sql like `select * from tbl where col1 > 100 and (col2 = 1 or col3 =2)`, the `(col2 = 1 or col3 =2)` predicate can actually filter lots of rows, but our index doesn't support this predicate. We can transform it to:
```
select * from tbl where  col1 > 100 and col2 = 1
union all
select * from tbl where col1 > 100 and col3 = 2 and (col2 !=1 or col2 is null);
```
In the second child, `col3 = 2` helps filter most rows and the other preidicate can based on its result do the rest filter job.

Why not make modifications on the BE? There are several reasons:
1. Modifying existing indexes and late materialization to support OR expressions is a complex work.
2. In clients' scenarios, predicates can be complex, and the selectivity and statistical information are closely related. It is more efficient for the FE to filter out predicates that meet the "good case" criteria and simplify them into a standardized form before handing them over to the BE for processing. 

### split strategy
1. Check if the scan data size is greater than `SCAN_OR_TO_UNION_THRESHOLD` with a default value of 1000000 rows * 50 bytes.
2. Split the conjunctions and determine if there already exists predicates with better selectivity. Only predicates that use a single column and have a selectRatio < 0.035 are considered to have good filtering selectivity.
3. If there are no predicates with good selectivity, further split the OR predicates. Check each element in the OR predicates to see if it qualifies as a predicate with good selectivity. If all elements qualify and the size of elements is samller than `SCAN_OR_TO_UNION_LIMIT`, split the OR predicate.

BTW, add a speicial process in clause with lots of values, we will split the values more than 1024 into some sets with most 1024 values helps use the index.

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.0
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 322 / 354 (90.96%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|7|11|63.64%|[443, 444, 463, 464]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NegateFilterShuttle.java|12|18|66.67%|[39, 52, 53, 63, 64, 72]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2164, 2165]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FilterSelectivityEvaluator.java|140|155|90.32%|[147, 167, 215, 260, 272, 296, 348, 357, 358, 360, 361, 363, 364, 365, 370]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|62|65|95.38%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java|84|86|97.67%|[71, 137]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 322 / 354 (90.96%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|7|11|63.64%|[443, 444, 463, 464]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NegateFilterShuttle.java|12|18|66.67%|[39, 52, 53, 63, 64, 72]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2164, 2165]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FilterSelectivityEvaluator.java|140|155|90.32%|[147, 167, 215, 260, 272, 296, 348, 357, 358, 360, 361, 363, 364, 365, 370]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|62|65|95.38%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java|84|86|97.67%|[71, 137]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 321 / 352 (91.19%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|7|11|63.64%|[443, 444, 463, 464]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NegateFilterShuttle.java|12|18|66.67%|[39, 52, 53, 63, 64, 72]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2164, 2165]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FilterSelectivityEvaluator.java|139|153|90.85%|[147, 167, 258, 270, 294, 346, 355, 356, 358, 359, 361, 362, 363, 368]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|62|65|95.38%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java|84|86|97.67%|[71, 137]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 321 / 352 (91.19%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|7|11|63.64%|[443, 444, 463, 464]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NegateFilterShuttle.java|12|18|66.67%|[39, 52, 53, 63, 64, 72]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2164, 2165]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FilterSelectivityEvaluator.java|139|153|90.85%|[147, 167, 258, 270, 294, 346, 355, 356, 358, 359, 361, 362, 363, 368]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|62|65|95.38%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java|84|86|97.67%|[71, 137]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java [hunk 1]
```diff
@@ -44,6 +44,9 @@ public final class LogicalOlapScanOperator extends LogicalScanOperator {
 
     private List<ScalarOperator> prunedPartitionPredicates;
 
+    // record if this scan is derived from SplitScanORToUnionRule
+    private boolean fromSplitOR;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java:48]: why need this mark?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 2]
```diff
@@ -1125,6 +1129,12 @@ public void setEnableParallelMerge(boolean enableParallelMerge) {
     @VarAttr(name = ENABLE_PLAN_VALIDATION, flag = VariableMgr.INVISIBLE)
     private boolean enablePlanValidation = true;
 
+    @VarAttr(name = SCAN_OR_TO_UNION_LIMIT, flag = VariableMgr.INVISIBLE)
+    private int scanOrToUnionLimit = 4;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1133]: How many scan operator for the or 4 Children

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    private List<OptExpression> transformImpl(OptExpression input, OptimizerContext context) {
+        LogicalOlapScanOperator scan = (LogicalOlapScanOperator) input.getOp();
+        long totalRowCount = StatisticsCalcUtils.getTableRowCount(scan.getTable(), scan);
+        Statistics.Builder builder = StatisticsCalcUtils.estimateScanColumns(scan.getTable(),
+                scan.getColRefToColumnMetaMap());
+        Statistics statistics = builder.setOutputRowCount(totalRowCount).build();
+
+        if (statistics.getComputeSize() <= context.getSessionVariable().getScanOrToUnionThreshold()) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java:84]: should consider no statistics case

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java (hunk 1 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java:48]: It's for the pattern check fast return. To aviod a the new scan child enter the rule again.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 2 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1133]: One scan can split to at most 4 union.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java (hunk 3 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java:84]: The `FilterSelectivityEvaluator` sets the filter to `NON_SELECTIVITY` when statistic is unknown, so no need to process it here.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 325 / 361 (90.03%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NegateFilterShuttle.java|16|27|59.26%|[42, 43, 44, 45, 47, 53, 70, 71, 81, 82, 90]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|7|11|63.64%|[443, 444, 463, 464]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2164, 2165]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FilterSelectivityEvaluator.java|139|153|90.85%|[147, 167, 258, 270, 294, 346, 355, 356, 358, 359, 361, 362, 363, 368]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|62|65|95.38%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java|84|86|97.67%|[71, 137]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java (hunk 3 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java:84]: Further confirm, what you mean is, if no statistics, will we spilt the or predicate to union?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+import com.starrocks.sql.optimizer.statistics.Statistics;
+import com.starrocks.sql.optimizer.statistics.StatisticsCalcUtils;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class SplitScanORToUnionRule extends TransformationRule {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java:42]: Add some comments and diagrams

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 325 / 361 (90.03%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/NegateFilterShuttle.java|16|27|59.26%|[42, 43, 44, 45, 47, 53, 70, 71, 81, 82, 90]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|7|11|63.64%|[443, 444, 463, 464]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2164, 2165]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/FilterSelectivityEvaluator.java|139|153|90.85%|[147, 167, 258, 270, 294, 346, 355, 356, 358, 359, 361, 362, 363, 368]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|62|65|95.38%|[95, 96, 97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitScanORToUnionRule.java|84|86|97.67%|[80, 146]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|2|2|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**